File: 28. Max Average Subarray II.py
# Topic: Max Average Subarray II (binary search + prefix sums)

def check(nums, k, mid):
    n = len(nums)
    prefix = [0.0] * (n+1)
    for i in range(n):
        prefix[i+1] = prefix[i] + nums[i] - mid
    min_pref = 0.0
    for i in range(k, n+1):
        if prefix[i] - min_pref >= 0:
            return True
        min_pref = min(min_pref, prefix[i-k+1])
    return False

def find_max_average(nums, k):
    low, high = min(nums), max(nums)
    for _ in range(40):
        mid = (low + high) / 2.0
        if check(nums, k, mid): low = mid
        else: high = mid
    return low

if __name__ == '__main__':
    nums = [1,12,-5,-6,50,3]
    k = 4
    print('Max avg:', round(find_max_average(nums,k),6))

# Time Complexity: O(n log(range/eps))
# Space Complexity: O(n)

File: 28. Max Average Subarray II.cpp
// Topic: Max Average Subarray II (find maximum average for subarray length >= k)
#include <bits/stdc++.h>
using namespace std;

bool check(const vector<int>& nums, int k, double mid)
{
    int n = nums.size();
    vector<double> prefix(n + 1, 0);
    for (int i = 0; i < n; ++i) prefix[i + 1] = prefix[i] + nums[i] - mid;
    double min_pref = 0;
    for (int i = k; i <= n; ++i)
    {
        if (prefix[i] - min_pref >= 0)
            return true;
        min_pref = min(min_pref, prefix[i - k + 1]);
    }
    return false;
}

double findMaxAverage(vector<int>& nums, int k)
{
    double low = *min_element(nums.begin(), nums.end());
    double high = *max_element(nums.begin(), nums.end());
    for (int iter = 0; iter < 40; ++iter)
    {
        double mid = (low + high) / 2;
        if (check(nums, k, mid))
            low = mid;
        else
            high = mid;
    }
    return low;
}

int main()
{
    vector<int> nums = {1, 12, -5, -6, 50, 3};
    int k = 4;
    cout.setf(std::ios::fixed);
    cout << setprecision(6);
    cout << "Max avg: " << findMaxAverage(nums, k) << '\n';
    return 0;
}

// Time Complexity: O(n log(range/eps))
// Space Complexity: O(n)
